File: game/scripts/vscripts/modifiers/modifier_sapphire.lua
modifier_sapphire = class({})

function modifier_sapphire:OnCreated()
    if IsServer() then
        self.efx_index = ParticleManager:CreateParticle("particles/generic_gameplay/rune_doubledamage_owner.vpcf", PATTACH_ABSORIGIN_FOLLOW, self:GetParent())
    end
end

function modifier_sapphire:OnDestroy()
    if IsServer() then
        ParticleManager:DestroyParticle(self.efx_index, false)
        ParticleManager:ReleaseParticleIndex(self.efx_index)
    end
end

function modifier_sapphire:GetStatusLabel() return "Sapphire" end
function modifier_sapphire:GetStatusPriority() return 2 end
function modifier_sapphire:GetStatusStyle() return "Sapphire" end

if IsClient() then require("wrappers/modifiers") end
Modifiers.Status(modifier_sapphire)
Modifiers.Shield(modifier_sapphire)

File: src/vscripts/modifiers/gems/modifier_sapphire.ts
import { registerModifier } from "../../lib/dota_ts_adapter";
import { ModifierShield } from "../modifier_shield";

@registerModifier("modifier_sapphire")
export class ModifierSapphire extends ModifierShield {
    particleId!: ParticleID;

    OnCreated(params: { damageBlock: number }) {
        super.OnCreated(params);

        if (IsServer()) {
            this.particleId = ParticleManager.CreateParticle(
                "particles/generic_gameplay/rune_doubledamage_owner.vpcf",
                ParticleAttachment.ABSORIGIN_FOLLOW,
                this.parent
            );
        }
    }

    OnDestroy() {
        super.OnDestroy();

        if (IsServer()) {
            ParticleManager.DestroyParticle(this.particleId, false);
            ParticleManager.ReleaseParticleIndex(this.particleId);
        }
    }

    GetTexture() {
        return "modifier_sapphire";
    }

    DeclareFunctions() {
        return [ModifierFunction.TOOLTIP];
    }

    OnTooltip() {
        return this.GetStackCount();
    }

    // GetStatusLabel() {
    //     return "Sapphire";
    // }
    // GetStatusPriority() {
    //     return 2;
    // }
    // GetStatusStyle() {
    //     return "Sapphire";
    // }
}

// Modifiers.Status(modifier_sapphire)
// Modifiers.Shield(modifier_sapphire)
